Add ExpandableText component to ui-components

jocrl opened the pull request:
This commit adds an ExpandableText component to ui-components.

<!--
  Please add a title to your Pull Request with the package the changes
  occured in.
-->
# ui-components // Add ExpandableText component

Collapsed:
<img width="214" alt="image" src="https://user-images.githubusercontent.com/91907326/171226111-15484ab7-0281-412d-8f5a-c19e2579f6a2.png">

Expanded:
<img width="209" alt="image" src="https://user-images.githubusercontent.com/91907326/171226143-22cbb41a-fa15-4732-98a7-a84911d0b0d6.png">

Does not exceed length:
<img width="213" alt="image" src="https://user-images.githubusercontent.com/91907326/171226187-38d669d3-01f7-4008-9ad6-b9af0b4623c0.png">

<!--
  Please describe your changes by describing motivation, linking to issues,
  and adding screenshots or gifs to illustrate your changes for reviewers
 -->
This component was added for https://github.com/cockroachdb/cockroach/pull/81180#issuecomment-1134899951, for the Job details error table.
It is being added in this repo per [this conclusion](https://cockroachlabs.slack.com/archives/C3T5LG9R8/p1645029579140249?thread_ts=1644965483.606339&cid=C3T5LG9R8):

> So, tl;dr for using db-console/cluster-ui components in managed-service:
> - For components (e.g. SQL Activity pages) that are specific to a database version and its APIs, they should live in cluster-ui and be imported in managed-service using componentFactory. This does a dynamic import that allows cluster-ui to be bundled separately from the main managed-service bundle
> - For generic components (e.g. Tooltips, SqlBox) that are not tied to a specific database version, they should ideally be moved to https://github.com/cockroachdb/ui. This allows them to be shared across db-console, cluster-ui, and managed-service
>   - As a fallback alternative, the code for these components can be duplicated in the managed-service repo, at the cost of code duplication

#### Checklist
- [x] I have written or updated test for the changes I made
- [ ] I have updated the README of the package I'm working in to reflect my changes
- [x] I have added or updated Storybook if appropriate for my changes

Annebirzin commented:
Nice! Can we just confirm it's using out core blue `#0055FF` ?

jocrl commented:
@Annebirzin updated, thanks for the catch!

maryliag reviewed (approved):
LGTM!

sjbarag commented on packages/ui-components/src/ExpandableText/ExpandableText.tsx:
```diff
@@ -0,0 +1,46 @@
+    <>
+      {!isExpanded && (
+        <>
+          {text.slice(0, characterLimit) + "..."}{" "}
```
nit: any chance we could use an ellipsis character (…) instead of three periods?

sjbarag reviewed (changes requested):
The example shown in https://github.com/cockroachdb/cockroach/pull/81180#issuecomment-1134899951 is a multi-line element, but the `ExpandableText` element here is based on pure character count.  Since proportional fonts will produce different widths of text with the same number of characters (think "iii" vs "WWW"), I'm not sure how we'll be able to accurately limit a component to two rendered lines this way 🤔 

Is this something we could accomplish by adding/removing `text-overflow: ellipsis` (https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow) to the text we want expanded, rather than using character counts?

sjbarag commented:
> The example shown in [cockroachdb/cockroach#81180 (comment)](https://github.com/cockroachdb/cockroach/pull/81180#issuecomment-1134899951) is a multi-line element, but the `ExpandableText` element here is based on pure character count. Since proportional fonts will produce different widths of text with the same number of characters (think "iii" vs "WWW"), I'm not sure how we'll be able to accurately limit a component to two rendered lines this way 🤔
> 
> Is this something we could accomplish by adding/removing `text-overflow: ellipsis` (https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow) to the text we want expanded, rather than using character counts?

Right, so CSS `text-overflow: ellipsis` only supports single-line overflows.  Using multiple lines would require either `-webkit-line-clamp` (which won't work on IE11: https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp) or a JS solution like https://www.npmjs.com/package/react-lines-ellipsis or the version provided by `antd` (https://ant.design/components/typography/#components-typography-demo-ellipsis; though IIRC we're trying to avoid antd in this repo?)

jocrl commented:
Ah, thanks Sean! I found antd's Expand, but hadn't found the ellipsis.

I initially put this in this repo to hopefully contribute to a cleaner unified UI state, but if there's concerns about having antd in this repo I'll ask that question on #frontend-learners and perhaps just put this component in cluster-ui for now + an issue for the future. I'll want to wrap the antd component in our own in order to have the Show More/Show Less, so it's not just an issue of directly using an antd component.

nathanstilwell commented on packages/ui-components/src/ExpandableText/ExpandableText.module.scss:
```diff
@@ -0,0 +1,10 @@
+@import "../styles/tokens.scss";
+
+.crl-anchor {
```
Since this is a CSS module, you don't need the `crl-` prefix for this classname.

jocrl commented:
Modified the implementation to truncate by lines instead of character count. Removed the test, since RTL doesn't seem to correctly take the fixed width into account when rendering.
Also removed the `crl-` CSS prefix.

Thanks for the feedback; PTAL!

nathanstilwell reviewed (commented):
A few comments ...

sjbarag commented on packages/ui-components/src/ExpandableText/ExpandableText.tsx:
```diff
@@ -0,0 +1,59 @@
+}: ExpandableTextProps): React.ReactElement {
+  const [isExpanded, setIsExpanded] = useState(false);
+
+  useEffect(() => {
```
> but wouldn't that happen anyway on a re-render (since the default state is false)?

IIRC the default value is only applied when the component is mounted, so an effect is necessary here I think!

jocrl commented on packages/ui-components/src/ExpandableText/ExpandableText.tsx:
```diff
@@ -0,0 +1,60 @@
+        <LinesEllipsis
+          text={text}
+          maxLine={maxLine}
+          ellipsis={
```
Thanks everyone!

@nathanstilwell, yeah the [type definition](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-lines-ellipsis/index.d.ts#L14) for the `ellipsis` prop is not up-to-date with the docs, and this is generating a typescript error. What do you mean by "disabling eslint anyway"? It looks like the CI runs `tsc`, which is failing (and is a helpful guard to have always passing?)

nathanstilwell replied:
Sorry, I was thinking about this wrong earlier (I mixed up `eslint-ignore-next-line` with `@ts-ignore`). Ultimately while it may not be better than `as any` (and disabling the linter), you could just cast the `ellipsis` block as a `string`.

```jsx
ellipsis={
  (
    <>
      {"… "}
      <a
        className={cx("crl-anchor")}
        onClick={() => {
          setIsExpanded(true);
        }}
      >
        Show More
      </a>
    </>
  ) as unknown as string
}

```
